src/cmap/connection.ts

```diff
@@ -757,3 +769,534 @@ function write(
+      completeCleanup();
+      return;
+    }
+
+    if (!this.socket.writableEnded) {
+      this.socket.end(() => {
+        this.socket.destroy();
+        completeCleanup();
+      });
+    } else {
+      completeCleanup();
+    }
+  }
+
+  command(
+    ns: MongoDBNamespace,
+    command: Document,
+    options: CommandOptions | undefined,
+    callback: Callback
+  ): void {
+    let cmd = { ...command };
+
+    const readPreference = getReadPreference(options);
+    const session = options?.session;
+
+    let clusterTime = this.clusterTime;
+
+    if (this.serverApi) {
+      const { version, strict, deprecationErrors } = this.serverApi;
+      cmd.apiVersion = version;
+      if (strict != null) cmd.apiStrict = strict;
+      if (deprecationErrors != null) cmd.apiDeprecationErrors = deprecationErrors;
+    }
+
+    if (this.hasSessionSupport && session) {
+      if (
+        session.clusterTime &&
+        clusterTime &&
+        session.clusterTime.clusterTime.greaterThan(clusterTime.clusterTime)
+      ) {
+        clusterTime = session.clusterTime;
+      }
+
+      const err = applySession(session, cmd, options);
+      if (err) {
+        return callback(err);
+      }
+    } else if (session?.explicit) {
+      return callback(new MongoCompatibilityError('Current topology does not support sessions'));
+    }
+
+    // if we have a known cluster time, gossip it
+    if (clusterTime) {
+      cmd.$clusterTime = clusterTime;
+    }
+
+    if (
+      // @ts-expect-error ModernConnections cannot be passed as connections
+      isSharded(this) &&
+      !this.supportsOpMsg &&
+      readPreference &&
+      readPreference.mode !== 'primary'
+    ) {
+      cmd = {
+        $query: cmd,
+        $readPreference: readPreference.toJSON()
+      };
+    }
+
+    const commandOptions: Document = Object.assign(
+      {
+        numberToSkip: 0,
+        numberToReturn: -1,
+        checkKeys: false,
+        // This value is not overridable
+        secondaryOk: readPreference.secondaryOk()
+      },
+      options
+    );
+
+    const message = this.supportsOpMsg
+      ? new OpMsgRequest(ns.db, cmd, commandOptions)
+      : new OpQueryRequest(ns.db, cmd, commandOptions);
+
+    try {
+      write(this as any as Connection, message, commandOptions, callback);
+    } catch (err) {
+      callback(err);
+    }
+  }
+}
+
+const kDefaultMaxBsonMessageSize = 1024 * 1024 * 16 * 4;
+
+/**
+ * @internal
+ *
+ * This helper reads chucks of data out of a socket and buffers them until it has received a
+ * full wire protocol message.
+ *
+ * By itself, produces an infinite async generator of wire protocol messages and consumers must end
+ * the stream by calling `return` on the generator.
+ *
+ * Note that `for-await` loops call `return` automatically when the loop is exited.
+ */
+export async function* readWireProtocolMessages(
+  connection: ModernConnection
+): AsyncGenerator<Buffer> {
+  const bufferPool = new BufferPool();
```

Comment:
Was this supposed to be the pool on the connection?

Reply:
no

Reply:
Why not, I'm surprised we need to make a new pool on every read? I see the one on the class is removed now, was there an issue with preserving it across reads?

Reply:
It's a leaky abstraction.  If `readWireProtocolMessages ` is the single abstraction we use to read full wire protocol messages from the socket, we shouldn't store the state it needs on the connection where it's not needed.

If this proves to be an impact on performance, we can reconsider